berserk/clients/explorer.py

```diff
@@ -0,0 +1,146 @@
+from __future__ import annotations
+
+from typing import cast, List, Literal, TypedDict
+import requests
+
+from ..exceptions import BerserkError
+from .base import BaseClient
+
+EXPLORER_URL = "https://explorer.lichess.ovh"
+
+OpeningExplorerVariant = Literal[
+    "standard",
+    "chess960",
+    "crazyhouse",
+    "antichess",
+    "atomic",
+    "horde",
+    "kingOfTheHill",
+    "racingKings",
+    "threeCheck",
+    "fromPosition",
+]
+
+
+Speed = Literal[
+    "ultraBullet", "bullet", "blitz", "rapid", "classical", "correspondence"
+]
+
+OpeningExplorerRating = Literal[
+    "0", "1000", "1200", "1400", "1600", "1800", "2000", "2200", "2500"
+]
+
+
+class Opening(TypedDict):
+    # The eco code of this opening
+    eco: str
+    # The name of this opening
+    name: str
+
+
+class Player(TypedDict):
+    # The name of the player
+    name: str
+    # The rating of the player during the game
+    rating: int
+
+
+class Game(TypedDict):
+    # The move in Universal Chess Interface notation
+    uci: str
+    # The id of the game
+    id: str
+    # The winer of the game. Draw if None
+    winner: Literal["white"] | Literal["black"] | None
+    # The speed of the game
+    speed: Speed
+    # The type of game
+    mode: Literal["rated"] | Literal["casual"]
+    # The black player
+    black: Player
+    # The white player
+    white: Player
+    # The year of the game
+    year: int
+    # The month and year of the game. For example "2023-06"
+    month: str
+
+
+class Move(TypedDict):
+    # The move in Universal Chess Interface notation
+    uci: str
+    # The move in algebraic notation
+    san: str
+    # The average rating of games in the position after this move
+    averageRating: str
+    # The number of white winners after this move
+    white: int
+    # The number of black winners after this move
+    black: int
+    # The number of draws after this move
+    draws: int
+    # The game where the move was played
+    game: Game | None
+
+
+class OpeningStatistic(TypedDict):
+    # Number of game won by white from this position
+    white: int
+    # Number of game won by black from this position
+    draws: int
+    # Number draws from this position
+    black: int
+    # Opening info of this position
+    opening: Opening | None
+    # The list of moves played by players from this position
+    moves: List[Move]
+    # recent games with this opening
+    recentGames: List[Game]
+    # top rating games with this opening
+    topGames: List[Game]
+
+
+class Explorer(BaseClient):
+    """Openings explorer endpoints."""
+
+    def __init__(self, session: requests.Session, explorer_url: str | None = None):
+        super().__init__(session, explorer_url or EXPLORER_URL)
+
+    def get_lichess_games(
+        self,
+        variant: OpeningExplorerVariant = "standard",
+        position: str | None = None,
+        play: str | None = None,
+        speeds: List[Speed] | None = None,
+        ratings: List[OpeningExplorerRating] = [],
```

Comment:
I did this to avoid having the type ``List[OpeningExplorerRating] | None``.
With this union typing, one can use the method like this:

```python
client.opening_explorer.get_lichess_games(ratings=None)
```

For me this seems weird as it is by definition a list of ratings.

For the sake of security I can add this line to force de re-instanciation of an empty list:

```python
ratings = ratings or []
```

I can also switch to ``None`` as default value.

As you wan't :)

Reply:
You can interpret it as "ratings filter = None" i.e. don't filter by ratings, just get all. It's not 100% perfect but this is a pretty standard way to handle default params in Python and even in the same endpoint you used it for the `speeds` which is also a list. So for consistency, I'd do it the same way here. And I guess, it's definitely better than using an empty list to mean "all ratings".